client/src/components/dashboard/DashboardViewer.jsx

```diff
@@ -0,0 +1,144 @@
+import React, { Component, PropTypes } from 'react';
+import DashboardViewerItem from './DashboardViewerItem';
+
+require('../../styles/DashboardViewer.scss');
+
+const viewportLimits = [
+  {
+    limit: 720,
+    name: 'small',
+  },
+  {
+    limit: 1024,
+    name: 'medium',
+  },
+  {
+    limit: Infinity,
+    name: 'large',
+  },
+];
+
+const getArrayFromObject = object => Object.keys(object).map(key => object[key]);
+
+export default class DashboardViewer extends Component {
+  constructor() {
+    super();
+    this.state = {
+      canvasWidth: 1080,
+      viewportType: 'large',
+    };
+
+    this.getItemFromProps = this.getItemFromProps.bind(this);
+    this.handleResize = this.handleResize.bind(this);
+  }
+
+  componentDidMount() {
+    this.handleResize();
+    window.addEventListener('resize', this.handleResize);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.handleResize);
+  }
+
+  getItemFromProps(item) {
+    switch (item.type) {
+      case 'text':
+        return item;
+
+      case 'visualisation': {
+        const output = Object.assign({}, item);
+
+        output.visualisation = this.props.visualisations[item.id];
+        return output;
+      }
+
+      default:
+        throw new Error(`Unknown item.type ${item.type} supplied to getItemFromProps()`);
+    }
+  }
+
+  handleResize() {
+    const width = this.refs.DashboardViewer.clientWidth;
+    let viewport;
+
+    for (let i = 0; i < viewportLimits.length; i++) {
+      const entry = viewportLimits[i];
+
+      if (width < entry.limit) {
+        viewport = entry.name;
+        break;
+      }
+    }
+
+    this.setState({
+      canvasWidth: width,
+      viewportType: viewport,
+    });
+  }
+
+  render() {
+    const { dashboard, datasets } = this.props;
+    const layout = dashboard.layout;
+
+    const sortFunc = (a, b) => {
+      const ay = layout[a.id].y;
+      const by = layout[b.id].y;
+      const ax = layout[a.id].x;
+      const bx = layout[b.id].x;
+
+      if (ay < by) {
+        return -1;
+      } else if (ay > by) {
+        return 1;
+      } else if (ax < bx) {
+        return -1;
+      } else if (ax > bx) {
+        return 1;
+      }
+      return 0;
+    };
+
+    const sortedDashboard = getArrayFromObject(dashboard.entities).sort(sortFunc);
```

Comment:
Since [Array.prototype.sort](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) is destructive, should we slice(0) first?

```
> foo = [2,1,3];
[2, 1, 3]
> foo.sort();
[1, 2, 3]
> foo
[1, 2, 3] // Note, foo changed
> foo = [2,1,3];
[2, 1, 3]
> foo.slice(0).sort();
[1, 2, 3]
> foo
[2, 1, 3]
```